### PR TITLE
[unitaryhack] Adds a new ZZ operation

### PR DESCRIPTION
**Pull request by Miru19**
**Context:** We would like to define a new diagonal operation called ZZ, as defined in the PennyLane-Braket plugin.

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:** #1148

### CODE REVIEW

**Comment by josh146** [comment]
Thanks for the contribution @Miru19! Let us know if you have any questions, or simply let us know in a comment when this is ready for review.

**Comment by Miru19** [comment]
@antalszava please let me know if I may set this PR as ready for review. Thanks!

**Review by antalszava — CHANGES_REQUESTED** [review]
Hi @miru19, thank you for submitting this! :tada:

1. Left a couple of suggestions for formatting that will help with rending the docs. The rendered documentation can be checked by clicking on the "Details" button next to the `docs/readthedocs.org:pennylane` check in the list of checks at the bottom of the Pull Request. This will redirect to a newly built version of the documentation, where searching for `IsingZZ` will find the documentation page for the new operation. Worth checking that this is fine.

2. It would be good to add a test for the gradient of this gate to check the gradient recipe.

This can be easily done by parametrizing an [existing test](https://github.com/PennyLaneAI/pennylane/blob/master/tests/ops/test_qubit_ops.py#L1730) we have for the ZZ Pauli Word too:
By modifying the signature of the test:
```python
    @pytest.mark.parametrize("angle", np.linspace(0, 2 * np.pi, 7))
    @pytest.mark.parametrize("pauli_word", ["XX", "ZZ"])
    def test_differentiability(self, angle, pauli_word, tol):
```
And using the new `pauli_word` argument within the body of the test:
```python
    qml.PauliRot(theta, pauli_word, wires=[0, 1])
```
Please ensure locally that with these modifications the test indeed works as expected.

Let us know if you have any questions! :slightly_smiling_face:

**Comment by antalszava** [comment]
Hi @Miru19, it seems that the formatting check is failing. To get it to pass, you will need to run `black` on both the source code and the tests directory (the commands are meant to be executed from the outer `pennylane` directory):
```bash
pip install black --upgrade
black -l 100 pennylane
black -l 100 tests
```

**Comment by Miru19** [comment]
@antalszava thank you! I'll solve it and I'll push the changes.
Thank you so much for letting me contribute to this project and for help!

File: pennylane/ops/qubit.py [hunk 1]
```diff
@@ -1777,7 +1777,11 @@ class IsingXX(Operation):
 
     * Number of wires: 2
     * Number of parameters: 1
-    * Gradient recipe: :math:`\frac{d}{d \phi} \mathtt{XX}(\phi) = \frac{1}{2} \left[ \mathtt{XX}(\phi + \pi / 2) + \mathtt{XX}(\phi - \pi / 2) \right]`
+    * Gradient recipe:
+    
+    .. math::
+        \frac{d}{d \phi} \mathtt{ZZ}(\phi)
+        = \frac{1}{2} \left[ \mathtt{ZZ}(\phi + \pi / 2) + \mathtt{ZZ}(\phi - \pi / 2) \right]
```
Comment (antalszava) [inline pennylane/ops/qubit.py:1784]:
```suggestion
* Gradient recipe: :math:`\frac{d}{d \phi} \mathtt{XX}(\phi) = \frac{1}{2} \left[ \mathtt{XX}(\phi + \pi / 2) + \mathtt{XX}(\phi - \pi / 2) \right]`
```

File: pennylane/ops/qubit.py [hunk 2]
```diff
@@ -1777,7 +1777,7 @@ class IsingXX(Operation):
 
     * Number of wires: 2
     * Number of parameters: 1
-    * Gradient recipe: :math:`\frac{d}{d \phi} \mathtt{XX}(\phi) = \frac{1}{2} \left[ \mathtt{XX}(\phi + \pi / 2) + \mathtt{XX}(\phi - \pi / 2) \right]`
+* Gradient recipe: :math:`\frac{d}{d \phi} \mathtt{XX}(\phi) = \frac{1}{2} \left[ \mathtt{XX}(\phi + \pi / 2) + \mathtt{XX}(\phi - \pi / 2) \right]`
```
Comment (antalszava) [inline pennylane/ops/qubit.py:1780]:
```suggestion
    * Gradient recipe: :math:`\frac{d}{d \phi} \mathtt{XX}(\phi) = \frac{1}{2} \left[ \mathtt{XX}(\phi + \pi / 2) + \mathtt{XX}(\phi - \pi / 2) \right]`
```

File: pennylane/devices/tests/test_gates.py [hunk 3]
```diff
@@ -180,7 +190,8 @@
 # list of all non-parametrized two-qubit gates
 two_qubit = [(qml.CNOT, CNOT), (qml.SWAP, SWAP), (qml.ISWAP, ISWAP), (qml.CZ, CZ), (qml.CY, CY)]
 # list of all parametrized two-qubit gates
-two_qubit_param = [(qml.CRX, crx), (qml.CRY, cry), (qml.CRZ, crz), (qml.IsingXX, IsingXX)]
+two_qubit_param = [(qml.CRX, crx), (qml.CRY, cry), (qml.CRZ, crz), (qml.IsingXX, IsingXX),
+                   (qml.IsingZZ, IsingZZ)]
```
Comment (antalszava) [inline pennylane/devices/tests/test_gates.py:194]:
```suggestion
two_qubit_param = [(qml.CRX, crx), (qml.CRY, cry), (qml.CRZ, crz), (qml.IsingXX, IsingXX), (qml.IsingZZ, IsingZZ)]
```

**Comment by antalszava** [comment]
Hi @Miru19, when merging in the `master` branch and resolving conflicts, the formatting of a file seems to have gone off, could you reformat the test files please? :slightly_smiling_face: 

Also, could you add the operation to [the `doc/introduction/operations.rst` file](https://github.com/PennyLaneAI/pennylane/blame/master/doc/introduction/operations.rst#L74) too so that it appears in the documentation?

**Comment by Miru19** [comment]
Hello @antalszava.
Sure, I'll solve it!
Thank you!

**Comment by Miru19** [comment]
Done

File: pennylane/ops/qubit.py [hunk 4]
```diff
@@ -1812,6 +1812,58 @@ def adjoint(self):
...
+        ]
+
+    @classmethod
+    def _matrix(cls, *params):
+        phi = params[0]
+        pos_phase = np.exp(1.0j * phi / 2)
+        neg_phase = np.exp(-1.0j * phi / 2)
+        return np.array(
+            [[neg_phase, 0, 0, 0], [0, pos_phase, 0, 0], [0, 0, pos_phase, 0], [0, 0, 0, neg_phase]]
+        )
```
Comment (antalszava) [inline pennylane/ops/qubit.py:1860]: (Minor suggestion): the way in the Braket plugin is nicely brief, perhaps good to have it here too?
```suggestion
        return np.diag([neg_phase, pos_phase, pos_phase, neg_phase])
```

Comment (Miru19) [reply pennylane/ops/qubit.py]: Done

**Comment by Miru19** [comment]
Thank you, too!
Thank you for the opportunity!

**Comment by crazy4pi314** [comment]
Hi @Miru19! 
Congrats on the contribution 🎉
This issue is part of [Unitary Hack](http://hack2021.unitary.fund) where you can register and get digital and physical swag for your contribution! If you want some cool stuff sent your way, [sign up here](https://airtable.com/embed/shrTG20PLC5cjVIPr)

**Comment by crazy4pi314** [comment]
@Miru19 I am going through and working on sending out swag, can you [sign up here](https://airtable.com/embed/shrTG20PLC5cjVIPr) so I send you your prizes?

**Pull request merged**